Notebooks/Table Header Detection.ipynb

```diff
@@ -567,44 +582,124 @@
     "\n",
-    "def extractCellFeatures(cell, startRowIndex):\n",
-    "    cellStyle = getCellStyle(cell)\n",
-    "    boolCellFeatures = {\n",
-    "        **extractLayoutFeatures(cell, cellStyle),\n",
-    "        **extractStyleFeatures(cell, cellStyle),\n",
-    "        **extractValueFeatures(cell)\n",
-    "    }\n",
-    "    intCellFeatures = mapDictBoolValuesToInt(boolCellFeatures)\n",
-    "    # intCellFeatures = applyColSpanFactor(intCellFeatures, getColSpan(cell))\n",
-    "    cellFeatures = []\n",
-    "    rowSpan = getRowSpan(cell)\n",
-    "    for rowIndex in range(startRowIndex, startRowIndex + rowSpan):\n",
-    "        featureCopy = dict(intCellFeatures)\n",
-    "        featureCopy['row'] = rowIndex\n",
-    "        featureCopy['colCount'] = 1\n",
-    "        cellFeatures.append(featureCopy)\n",
-    "    return cellFeatures\n",
-    "\n",
     "def merge(featuresA, featuresB):\n",
     "    return { k: featuresA.get(k, 0) + featuresB.get(k, 0) for k in set(featuresA) | set(featuresB) }\n",
     "\n",
     "def stringifyDictKeys(dictionary):\n",
     "    return { str(key): value for key, value in dictionary.items() }\n",
     "\n",
-    "def extractFeatures(table):\n",
+    "def numNormalisedCols(row):\n",
+    "    numCols = 0\n",
+    "    for cell in row.children:\n",
+    "        if type(cell) is Tag:\n",
+    "            numCols += getColSpan(cell)\n",
+    "    return numCols\n",
+    "            \n",
+    "def getSimilarity(feature, cell, neighbour, suffix):\n",
+    "    similarity = {}\n",
+    "    similarity[feature + 'A' + suffix] = cell[feature] and neighbour[feature]\n",
+    "    similarity[feature + 'B' + suffix] = cell[feature] and not neighbour[feature]\n",
+    "    return similarity\n",
+    "    \n",
+    "def extractSimilarityFeatures(cell, neighbour, suffix):\n",
+    "    similarityFeatures = {\n",
+    "        **getSimilarity('isMerged', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isCenterAligned', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isTHOrInTHead', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isBold', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isItalic', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isUnderlined', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isEmpty', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isNumeric', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isDate', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isShortText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isLongText', cell, neighbour, suffix),\n",
+    "        **getSimilarity('isTotal', cell, neighbour, suffix)        \n",
+    "    }\n",
+    "    return similarityFeatures    \n",
+    "    \n",
+    "def addSimilarityFeatures(normalizedFeatureTable):\n",
```

Comment:
This method seems to be rather long. Maybe we can split it by extraction the inner for-loop into an extra method?